File: kotlin/Lib/264.UglyNumberII.kt
// 264. Ugly Number II


class LeetCode264UglyNumberII {
    private data class PrimePointer(val `val`: Int, var pos: Int)
    fun nthUglyNumber(n: Int): Int {
        if (n == 1) return 1

        val dp = IntArray(n)
        dp[0] = 1

        val pps = arrayOf(
            PrimePointer(2, 0),
            PrimePointer(3, 0),
            PrimePointer(5, 0))

        for (i in 1 until n) {
            dp[i] = pps.map { p -> p.`val` * dp[p.pos] }.min()!!

            for (p in pps)
                if (dp[i] == p.`val` * dp[p.pos])
                    ++p.pos
        }

        return dp[n-1]
    }
}

File: csharp/src/264_UglyNumberII.cs
// LeetCode: 264.Ugly Number II

using System;
using System.Linq;

namespace LeetCode.Problem_264
{
	public class Solution
	{
		private class PrimePointer
		{
			public int Val;
			public int Pos;
		}
		public int NthUglyNumber(int n)
		{
			var primes = new PrimePointer[]
			{
				new PrimePointer
				{
					Val = 2,
					Pos = 0,
				},
				new PrimePointer
				{
					Val = 3,
					Pos = 0,
				},
				new PrimePointer
				{
					Val = 5,
					Pos = 0,
				},
			};

			var table = new int[n];
			table[0] = 1;
			for (int i = 1; i < table.Length; ++i)
			{
				var minVal = int.MaxValue;
				foreach (var prime in primes)
					minVal = Math.Min(minVal, prime.Val * table[prime.Pos]);
				foreach (var prime in primes)
					if (minVal == prime.Val * table[prime.Pos])
						++prime.Pos;
				table[i] = minVal;
			}
			return table[n-1];
		}
	}
}
